src/raftstore/store/snap.rs

```diff
@@ -618,6 +620,1006 @@ mod v1 {
+                *size_track = size_track.saturating_sub(size_to_sub);
+            }
+            Ok(())
+        }
+
+        fn validate(&self) -> RaftStoreResult<()> {
+            let cf_size_checksums =
+                box_try!(decode_cf_size_checksums(self.meta_file.data.get_data()));
+            for (cf, expected_size, expected_checksum) in cf_size_checksums {
+                let mut checked = false;
+                for cf_file in &self.cf_files {
+                    if cf_file.cf != cf {
+                        continue;
+                    }
+                    let size = try!(get_file_size(&cf_file.path));
+                    if size != expected_size {
+                        return Err(box_err!("invalid snapshot file size {} for cf {}, expected \
+                                             {}",
+                                            size,
+                                            cf_file.cf,
+                                            expected_size));
+                    }
+                    let checksum = try!(calc_checksum(&cf_file.path));
+                    if checksum != expected_checksum {
+                        return Err(box_err!("invalid snapshot file checksum {} for cf {}, \
+                                             expected {}",
+                                            checksum,
+                                            cf_file.cf,
+                                            expected_checksum));
+                    }
+                    checked = true;
+                }
+                assert!(checked);
+            }
+            Ok(())
+        }
+
+        fn switch_to_cf_file(&mut self, cf: String) -> io::Result<()> {
+            let mut cf_found = false;
+            let mut index = 0;
+            for (i, f) in self.cf_files.iter().enumerate() {
+                if f.cf == cf {
+                    cf_found = true;
+                    index = i;
+                    break;
+                }
+            }
+            if !cf_found {
+                return Err(io::Error::new(ErrorKind::Other, format!("fail to find cf {}", cf)));
+            }
+            self.cf_index = index;
+            Ok(())
+        }
+
+        fn add_kv(&mut self, k: &[u8], v: &[u8]) -> io::Result<()> {
+            let mut cf_file = &mut self.cf_files[self.cf_index];
+            let mut writer = cf_file.sst_writer.as_mut().unwrap();
+            if let Err(e) = writer.add(k, v) {
+                return Err(io::Error::new(ErrorKind::Other, e));
+            }
+            cf_file.kv_count += 1;
+            Ok(())
+        }
+
+        fn save_cf_files(&mut self) -> io::Result<()> {
+            for cf_file in &mut self.cf_files {
+                if cf_file.kv_count == 0 {
+                    let _ = cf_file.sst_writer.take().unwrap();
+                } else {
+                    let mut writer = cf_file.sst_writer.take().unwrap();
+                    if let Err(e) = writer.finish() {
```

Comment:
It fails to cast return type from `Result<(), String>` to `io::Result<()>` using `try!` or `box_try!`.